pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -447,28 +442,62 @@ func newInboundPassthroughFilterChains(configgen *ConfigGeneratorImpl,
 			Push:             push,
 			ListenerProtocol: plugin.ListenerProtocolTCP,
 		}
+		var allChains []plugin.FilterChain
+		for _, p := range configgen.Plugins {
+			chains := p.OnInboundPassthroughFilterChains(in)
+			allChains = append(allChains, chains...)
+		}
+
+		if len(allChains) == 0 {
+			// Add one empty entry to the list if none of the plugins are interested in updating the filter chains.
+			allChains = []plugin.FilterChain{{}}
+		}
+		// Override the filter chain match to make sure the pass through filter chain captures the pass through traffic.
+		for i := range allChains {
+			chain := &allChains[i]
+			if chain.FilterChainMatch == nil {
+				chain.FilterChainMatch = &listener.FilterChainMatch{}
+			}
+			// Port : EMPTY to match all ports
```

Comment:
Can these be part of the `OnInboundPassthroughFilterChains` instead? I think the port and listener protocol can, and we can pass matchingIP via function args? In fact, would be nice to remove the whole loop if possible.

Reply:
OnInboundPassthroughFilterChains() is a callback for the plugin and the plugin could return a nil indicating it's not interested in updating the filter chain. For example when there is only alpha authN policy.

Also, updating the port/ip should be done out of the plugin as the logic doesn't belong to any of the specific plugins.

Reply:
I find it a bit odd that the OnInboundPassthroughFilterChains (may) create passthrough filter chain, but then need rewrite to make it fit. Plus, it is unlikely that all other plugins except authn generate new filter chain. Would it be cleaner to have a function to create the correct filter chain(s) in line 447, instead of calling plugin?

Reply:
OnInboundPassthroughFilterChains doesn't create the pass through filter chain. It's used to allow the plugin to create one (or more) basic filter chains needed as the template. The listener_builder.go update the basic filter chains (template) to make it the pass through filter chain.

It's hard to say if it's more cleaner to wrap this into a separate function because I basically followed the same patter of the OnInboundFilterChains which does pretty much the same thing (for normal filter chain) and is only used by the istio_authn filter. 

I guess the listener_builder.go and listener.go was trying to decouple plugin related logic out of the core flow. That's probably why the OnInboundFilterChains was introduced instead of making it a standalone function.

@howardjohn @rshriram @incfly Could you comment what's the preferred approach here? Thanks.

Reply:
The filter chains output of that function should be used only for passthrough filter chains, right? At least that's what the function name and the place where it is called indicate. I don't see where else this function should be called and used for non-passthrough filters. That's why I find it confusing as you have to overwrite the output to make it fit for passthrough filter chain.

Reply:
If the pilot owner is okay with a standalone function here, sure, I can change the PR to use it. Just be noted that the standalone function needs to directly call the function defined under the authN plugin.

Reply:
So is it true also that only  authn plugin implements none empty logic of OnPassthroughFilterChain? If so, might be better just doing things in builder code here.

Reply:
authn plugin being the only filter using OnPassthroughFilterChain() doesn't really justify to put it here, the other OnInboundFilterChain() is also used only by authn plugin but it's well isolated in the plugin instead of putting it here.

I'm trying to ask is it okay to couple plugin-related logic to this file (listener_filter.go) because previously I got clear review comments from @howardjohn that we should avoid doing this, that's also my understanding of the rational to have the OnInboundFilter() in the first place.

Reply:
I'm not proud to say OnInboundFilter is a great example :). Anyway, I have no objection to use the plugin flow if you want to keep that.

My original concern is the `OnInboundPassthroughFilterChains` seems very specific for inbound filter chain already, but you said that it is for generic filter chain (and thus need the for loop to overwrite the settings) is confused to me.